src/aics-image-viewer/components/ViewerStateProvider/index.tsx

```diff
@@ -105,6 +95,53 @@ const applyChangeToViewerSettings = <K extends keyof ViewerState>(
   }
 };
 
+/** Utility type to explicitly assert that one or more properties will *not* be defined on an object */
+type WithExplicitlyUndefined<K extends keyof any, T> = T & { [key in K]?: never };
+
+/** Set channel setting `key` on one or more channels specified by `index` to value `value`. */
+type ChannelSettingUniformUpdateAction<K extends keyof ChannelState> = {
+  index: number | number[];
+  key: K;
+  value: ChannelState[K];
+};
+/** Set the values of channel setting `key` for all channels from an array of values ordered by channel index */
+type ChannelSettingArrayUpdateAction<K extends keyof ChannelState> = {
+  key: K;
+  value: ChannelState[K][];
+};
+/** Initialize list of channel states */
+type ChannelSettingInitAction = {
+  value: ChannelState[];
+};
+
+type ChannelStateAction<K extends keyof ChannelState> =
+  | ChannelSettingUniformUpdateAction<K>
+  | WithExplicitlyUndefined<"index", ChannelSettingArrayUpdateAction<K>>
+  | WithExplicitlyUndefined<"index" | "key", ChannelSettingInitAction>;
+
+const channelSettingsReducer = <K extends keyof ChannelState>(
```

Comment:
I like how the function manages different actions based on the `key` and `index` properties, effectively covering various updates. 

nit: are there potential cases where the index could be out of bounds or the key might be invalid?

Reply:
TypeScript should keep us from ever using an invalid key, but it's definitely possible to try to index the channel settings array out of bounds. I added a check!